{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"collapsed_sections":[]},"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.6.9"}},"cells":[{"cell_type":"markdown","metadata":{"id":"cOKChH14H_pR"},"source":["# Day 1 ‚Äì Tutorial 1:¬†Introduction to Google Colab + Visualizing Molecular Structures in Google Colab using py3Dmol"]},{"cell_type":"markdown","source":["## What is Google Colab? ü§î"],"metadata":{"id":"Vu8C_1osg0B1"}},{"cell_type":"markdown","source":["Colaboratory, or \"Colab\" for short, allows you to write and execute Python in your browser, with \n","- Zero configuration required\n","- Free access to GPUs\n","- Easy sharing\n","\n","You can watch the [Introduction to Colab](https://www.youtube.com/watch?v=inN8seMm7UI) video recommended by Google Colab to learn more, or just get started below!"],"metadata":{"id":"CCS1laExg_Bv"}},{"cell_type":"markdown","source":["### **The central concept of Jupyter Notebooks on Google Colab: Text Cells and Code Cells**\n","\n","A Jupyter notebook is a list of cells. Cells contain either **explanatory text (TEXT)** or **executable code (CODE)** and its output.\n","\n","You can add new cells by using the **+ CODE** and **+ TEXT** buttons that show when you hover between cells. These buttons are also in the toolbar above the notebook, and they can be used to add a cell below the currently selected cell.\n","\n","You can also move a cell by selecting it and clicking **Cell Up** or **Cell Down** in the top toolbar."],"metadata":{"id":"ce0ProrvhNz5"}},{"cell_type":"markdown","source":["### **Text Cells**\n","On Google Colaboratory, the text cells are formatted using a simple markup language called **markdown**, based on [the original](https://daringfireball.net/projects/markdown/syntax) markdown project. To learn more, see the [markdown\n","guide](/notebooks/markdown_guide.ipynb) recommended by Google Colab.\n","\n","For example, this is a **text cell**. You can **double-click** this cell to edit it.\n"," "],"metadata":{"id":"2QJAk2lBiVdT"}},{"cell_type":"markdown","source":["#### **Text Formats**\n","\n","You can also use tags to format your text. The following are examples of markdown text formats. Each word/phrase is shown in the desired format, and the tags around it are those required to achieve each specific format.\n","\n","\\**italics*\\* or \\__italics__\n","\n","**\\*\\*bold\\*\\***\n","\n","\\~\\~~~strikethrough~~\\~\\~\n","\n","\\``monospace`\\`\n","\n","**Indentations:**\n","\n","No indent\n",">\\>One level of indentation\n",">>\\>\\>Two levels of indentation\n","\n","**An ordered list:**\n","1. 1\\. One\n","1. 1\\. Two\n","1. 1\\. Three\n","\n","**An unordered list:**\n","* \\* One\n","* \\* Two\n","* \\* Three"],"metadata":{"id":"YTo1XSdRiQTH"}},{"cell_type":"markdown","source":["#### **Math, Equations, and Tables**\n","\n","You can also add math to text cells using [$\\LaTeX$](http://www.latex-project.org/)\n","to be rendered by [MathJax](https://www.mathjax.org). Just place the statement\n","within a pair of **`$`** signs. For example `$\\sqrt{3x-1}+(1+x)^2$` becomes\n","$\\sqrt{3x-1}+(1+x)^2.$\n","\n","Also, if you double the **`$`** tags in your $\\LaTeX$ equations, you can set the contents off on its own centered line. For example, `$$y = 0.1 x$$` renders the following equation: $$y = 0.1 x$$\n","\n","Finally, you can also easily generate tables with data following these indications:\n","\n","```\n","First column name | Second column name\n","--- | ---\n","Row 1, Col 1 | Row 1, Col 2\n","Row 2, Col 1 | Row 2, Col 2\n","```\n","\n","becomes:\n","\n",">First column name | Second column name\n",">--- | ---\n",">Row 1, Col 1 | Row 1, Col 2\n",">Row 2, Col 1 | Row 2, Col 2\n","\n","Horizontal rule done with three dashes (\\-\\-\\-):\n"],"metadata":{"id":"eD3789_HimTM"}},{"cell_type":"markdown","source":["### **Code Cells**\n","\n","Below is a **code cell**. To execute the contents of a code cell, you first must connect to a hosted runtime by clicking on the **Connect** button located in the toolbar menu.\n","\n"," <img src='https://media.giphy.com/media/lRLBURv0hpcHqiraBI/giphy.gif'/> \n","\n","Once the toolbar button changes to **'Connected'**, click in the code cell below to select it and execute the contents in the following ways:\n","\n","* Click the **Play icon** in the left gutter of the cell;\n","* Type **Cmd/Ctrl+Enter** to run the cell in place;\n","* Type **Shift+Enter** to run the cell and move focus to the next cell (adding one if none exists); or\n","* Type **Alt+Enter** to run the cell and insert a new code cell immediately below it.\n","\n","There are additional options for running some or all cells in the **Runtime** menu."],"metadata":{"id":"vaepD6GXj6jQ"}},{"cell_type":"markdown","source":["#### **Shell on Google Colab**\n","\n","Google Colab can interpret typical shell commands that you might have used in the past, such as:\n","\n","```bash\n","cd: change directory\n","ls: list\n","mv: move\n","cp: copy\n","mkdir: make new directory\n","history: terminal history (i.e. history of the commands you have executed) help: command list & help\n","echo: outputs to terminal\n","expr: evaluate expression and outputs to terminal\n","wc: word count\n","cat: ‚Äúconcatenate‚Äù streams all input to the terminal\n","sed: ‚Äústream editor‚Äù edits input (most importantly substitute) and outputs to the terminal \n","grep: ‚ÄúGlobal regular expression print‚Äù, searches a given expression and outputs to terminal \n","awk: pattern scanning & processing language\n","```\n","\n","Let's test some of these commands in the code cells below:"],"metadata":{"id":"mBysejYdkdhW"}},{"cell_type":"code","source":["#List all files and directories with ls\n"],"metadata":{"id":"EUg01lMElSA0"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Print the working directory of Google Colab cloud linux instance with pwd\n"],"metadata":{"id":"xC32XWtblU3y"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Print 'Hello world' using echo\n"],"metadata":{"id":"nYSBnFjAlV5E"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["\n","üí° **HINT**: You can use the %%bash at the beginning of your cell code instead of the `!` approach to active automagic for the whole cell code. Try it below!"],"metadata":{"id":"czwtuYuRmqdj"}},{"cell_type":"code","source":["%%bash\n","#List all files and directories with ls\n","\n","#Print the working directory of Google Colab cloud linux instance with pwd\n","\n","#Print 'Hello world' using echo\n"],"metadata":{"id":"DQlejVQ8mtuI"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["#### **Shell-Related Magic Commands**\n","\n","If you play around with some of these shell commands for a while, you will notice that some of them do not lead to permanent changes in your runtime environment. For example, let's try to create a directory and then permanently enter into it:"],"metadata":{"id":"k0NNm7HzlcUf"}},{"cell_type":"code","source":["#Creating a directory\n","!mkdir newfolder\n","!echo \"folder created\"\n","!echo \" \"\n","#Changing into that directory\n","!cd newfolder\n","!echo \"attempting to move into /content/newfolder\"\n","!echo \" \"\n","#Checking in which directory we are now\n","!echo \"we are on the following directory:\"\n","!pwd\n","#Getting back into the content folder\n","!echo \" \"\n","!echo \"going back to the content folder\"\n","!cd /content\n","!echo \" \"\n","#Checking in which directory we are now\n","!echo \"we are on the following directory:\"\n","!pwd"],"metadata":{"id":"eC1VhSIZl3ne"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["The reason is that shell commands in the notebook are executed in a temporary subshell.\n","\n","If you'd like to change the working directory in a more enduring way, you can use the ``%cd`` magic command.\n","\n","This is known as an ``automagic`` function, and this behavior can be toggled with the ``%automagic`` magic function.\n","\n","Besides ``%cd``, other available shell-like magic functions are ``%cat``, ``%cp``, ``%env``, ``%ls``, ``%man``, ``%mkdir``, ``%more``, ``%mv``, ``%pwd``, ``%rm``, and ``%rmdir``, any of which can be used without the ``%`` sign if ``automagic`` is on."],"metadata":{"id":"FzwpFHD-nCfF"}},{"cell_type":"code","source":["#Creating a directory\n","!mkdir newfolder\n","!echo \"folder created\"\n","!echo \" \"\n","#Changing into that directory with an automagic command cd\n","%cd newfolder\n","!echo \"attempting to move into /content/newfolder\"\n","!echo \" \"\n","#Checking in which directory we are now\n","!echo \"we are on the following directory:\"\n","!pwd\n","#Getting back into the content folder with an automagic command cd\n","!echo \" \"\n","!echo \"going back to the content folder\"\n","%cd /content\n","!echo \" \"\n","#Checking in which directory we are now\n","!echo \"we are on the following directory:\"\n","!pwd"],"metadata":{"id":"8Al6haS_nduW"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["#### **Python**\n","\n","The following is an example of a very simple code cell using **python**, which is the programming option by default for Google Colab."],"metadata":{"id":"3N2MoKX6n7Fm"}},{"cell_type":"code","source":["W = 'Tryptophan'\n","C = 'Cysteine'\n","W,C"],"metadata":{"id":"Mf-9oVv_kbDz"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["We will not get into the details of python, because we will be using it during the tutorials and (hopefully) learn a bit about it."],"metadata":{"id":"miEXNvUmoG9j"}},{"cell_type":"markdown","source":["## Visualizing protein structures using py3Dmol"],"metadata":{"id":"pkasrrYmo3wf"}},{"cell_type":"markdown","metadata":{"id":"Xa3XrE34Z0hj"},"source":["### **Theoretical aspects**\n"]},{"cell_type":"markdown","metadata":{"id":"2L0-BgSfb1Ss"},"source":["A PDB file contains information of all the experimental conditions that allowed obtaining a given three-dimensional protein structure (contained in the PDB file), as well as information about the biological features of the protein and the overall features of the crystal structure (cell unit, dimensions, number of monomers, biological assembly, etc.). However, its most relevant body of corresponds to the **coordinates** (i.e. the three-dimensional positions) of the atoms of a structure of a given protein. This is, all PDB files contain the following minimal format:\n","\n","![alt text](https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/pdbformat_01.png)\n","\n","In which **ATOM** (or HETATM) indicates that the line contains information of atomic coordinates, a unique number for each atom (a list that is also referred to as index), its atom type, the residue name to which a given atom belongs to, the polypeptide chain in which this residue is located, the position or number of the residue in the primary sequence, and the cartesian atomic coordinates of each atom. Any molecule (protein or not) can be written in this format, as long as we have the cartesian coordinates for each of its atoms.\n","\n","As you may recall from the lecture, biological structures are gathered by different experimental evidences: **X-ray crystallography**, **Nuclear Magnetic Resonance** or, more recently, **Cryo-Electron microscopy**. This information is far from giving you exact atomic coordinates resolutions, thus measurements of how well defined is a given coordinate is also written in this format. Those columns, next to the coordinates, are the occupancy and temperature (or B) factors. The occupancy tells you if the atom is definitely in that position, or if an alternative configuration is also likely. The image in **Figure 1** presents this scenario in a protein structure solved by X-ray crystallography (PDB 6ANE; a PET hydrolase from _I. sakaiensis_ solved by our laboratory), where a given tryptophan residue cannot be univocally assigned to a single configuration (i.e. the electron density is equally fit in both configurations). In this case, this is interpreted as if this residue actually _occupies_ both configurations in the crystal, such that nearly half of the molecules in the crystal lattice take one of the conformers."]},{"cell_type":"markdown","metadata":{"id":"L5e8DTyic7a7"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/pdbformat_01.gif'/>\n","<figcaption>FIGURE 1. Occupancies of a tryptophan residue in a crystallographic structure (PDB ID 6ANE). In this case, the same atom has two sets of coordinates, one for each configuration. Hence, the occupancy of each of those coordinate lines is reduced to 0.50. In the general case, when a single atom (line of coordinates) has n possible configurations, its occupancy is reduced to 1/n.</figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"yvTSiBA8CWxe"},"source":["üí° **HINT:** The GIF in Figure 1 is the result of comparing the same tryptophan residue after structural superposition of two different chains collected in the same X-Ray crystallography experiment. This illustrates the need to carefully inspect a PDB structure before doing any computational study with it, because if there are differences in a side chain that is near the active site that may impact the results of subsequent studies, such as molecular docking "]},{"cell_type":"markdown","metadata":{"id":"t4GdLbOZeiPB"},"source":["On the other hand, the **_temperature_ factors** are a measurement of uncertainty in the positions assigned to each atom. In the previous case, occupancy was telling us that we had multiple stable configurations within our structure. Instead, temperature factors (also referred to as isotropic B-values) are measuring how likely it is to see that atom in that particular position. The larger this value, the higher the mobility around our defined position, hence higher the uncertainty. This value depends inversely with resolution: high resolution structures have nicely-detailed atomistic resolution, and we can see with little doubt were each atom should be positioned, while low resolution structures only provide us with coarse features of the molecular ensemble, and coordinates are much more difficult to assign **(Figure 2)**."]},{"cell_type":"markdown","metadata":{"id":"Me6lkOlRelvh"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/pdbformat_02.png'/>\n","<figcaption>FIGURE 2. Effect of resolution on electron-density maps. In the first case we can clearly see the positions of the heavy atoms, while at lower resolution these positions become uncertain.</figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"Px_0dHGKewXJ"},"source":["The two parameters just defined are complementary to each other, as the multiple stable conformers arising from uncomplete occupancy are not necessarily equally likely, so mutual information helps you decide which is the most likely **(Figure 3)**."]},{"cell_type":"markdown","metadata":{"id":"q-F_lQ46evf6"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/pdbformat_03.png'/>\n","<figcaption>FIGURE 3. Coordinate lines for the same atom (CB) in two distinct tryptophan configurations (ATRP, BTRP). As you may notice, the occupancy is half for each, yet the B-factor is lower for the first (A) configuration, hence more likely.</figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"U3qgbkATe4zM"},"source":["\n","With this, we have briefly covered the essentials of PDB files and structures. However, the goodness of a given protein structure deposited in the PDB is present in the webpage of the corresponding PDB identifier, as show in **Figure 4**."]},{"cell_type":"markdown","metadata":{"id":"wtOmHs8Be7N7"},"source":["<figure>\n","<center>\n","<img src='https://raw.githubusercontent.com/pb3lab/ibm3202/master/images/pdbformat_04.png'/>\n","<figcaption>FIGURE 4. Overall structure quality. This plot depends on the structure and resolution. It ranks from worst (red) to best (blue) structures at the same resolution.</figcaption></center>\n","</figure>"]},{"cell_type":"markdown","metadata":{"id":"75xQSoMVfBt8"},"source":["<p align = \"justify\">Several parameters allow you to assess the quality of an X-ray solved protein structure. \n","\n",">In **Rwork/Rfree**, the former tells you how well fitted are the atomic coordinates to the experimental data, while the latter is a ‚Äúcontrol‚Äù measurement with part of the data that was left out in calculating Rwork. The lower these values are, the better the fit of the solved structure to the electron density. However, be careful if Rwork << Rfree, as it is likely that the data is being overfitted to the experimental electron density. \n",">\n",">The **Clashscore** measures if atoms are unusually close to each other (number of clashes per every 1,000 atoms). \n",">\n",">The **Ramachandran outliers** measure if there are unusually stretched torsions in the œÜ- and œà-dihedral angles of the backbone. \n",">\n",">Amino acid sidechains also adopt preferred configurations (or rotamers), thus **Sidechain outliers** measure how many uncommon rotamers are present in a given structure. \n",">\n",">Finally, **RSRZ** is a measurement between the atomic model and the real-space data.\n","\n","Although the PDB file contains the coordinates of a given structure in text, you have already seen that most solved protein structures are typically shown as visual representations of such atomic coordinates. "]},{"cell_type":"markdown","metadata":{"id":"WsSl31k4f38T"},"source":["### **Experimental Overview**\n","\n","Here, we will use **py3DMol** (python package for dependency-free molecular visualization in iPython notebooks) to visualize and analyze a protein structure that we will download from the Protein Data Bank (PDB)"]},{"cell_type":"markdown","source":["## Part 0 - Downloading and Installing the required software"],"metadata":{"id":"LOmaeGkBqe-q"}},{"cell_type":"markdown","source":["Before we start, you must first **remember to start the hosted runtime in Google Colab**.\n","\n","Then, we must install several pieces of software to perform this tutorial. Namely:\n","- **biopython** for manipulation of the PDB files\n","- **py3Dmol** for visualization of the protein structure\n","\n","Both software packages can be installed using `pip`"],"metadata":{"id":"ShVBz50KqpsX"}},{"cell_type":"code","source":["#Installing py3Dmol using pip\n","!pip install py3Dmol\n","#Installing biopython using pip\n","!pip install biopython\n","#Importing py3Dmol for safety\n","import py3Dmol"],"metadata":{"id":"aumqkK5yq03R"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"k3Qc6YJ5gIGt"},"source":["## Part 1. Retrieving a structure from the Protein Data Bank using biopython"]},{"cell_type":"markdown","metadata":{"id":"DYejo5FEgjSZ"},"source":["You can retrieve PDB structures from its website (www.rcsb.org). Alternatively, you can directly use Google Colab and biopython to download a given PDB file with known accession code as shown below. Here, we are downloading the structure of *I. sakaiensis* PET hydrolase solved by our research group using X-ray crystallography: [Fecker T et al (2018) *Biophys J 114 (6), 1302-1312*](https://doi.org/10.1016/j.bpj.2018.02.005)\n","\n"]},{"cell_type":"code","metadata":{"id":"S2GfVk67gisv"},"source":["#Downloading the PDB files using biopython\n","import os\n","from Bio.PDB import *\n","pdbid = ['6ane']\n","pdbl = PDBList()\n","for s in pdbid:\n","  pdbl.retrieve_pdb_file(s, pdir='.', file_format =\"pdb\", overwrite=True)\n","  os.rename(\"pdb\"+s+\".ent\", s+\".pdb\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"afW0mstAhGWV"},"source":["If you open one of such files with a text editor, you will be able to notice all of the things that we previously discussed about PDB files in this tutorial. Briefly, you will find plenty of experiment-related information, including the parameters of extraction and solving the crystal structure, followed by the information of the coordinates of each atom that was assigned to a given electron density after the X-ray diffraction and structure solution procedure."]},{"cell_type":"markdown","source":["Let's check how many chains are contained in the PDB file and how many residues has each chain by using some shell coommands, like `grep`"],"metadata":{"id":"LlDrOhoVrzpn"}},{"cell_type":"code","metadata":{"id":"jFANtPwvF2GK"},"source":["#Do you want to know how many chains are contained in the PDB? grep can help you! \n","!echo \"Checking the total number of chains\"\n","!grep 'COMPND'.*'CHAIN' 6ane.pdb\n","!echo \" \"\n","#How many residues has each chain? Now we can use awk!\n","!echo \"Checking the total number of residues for chain A\"\n","!awk '$1==\"ATOM\" && $3==\"CA\" && $5==\"A\" {print $0}' 6ane.pdb | wc -l"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"N4K33OukF-Yu"},"source":["As the structure of _I. sakaiensis_ PETase has 3 different chains, we will use biopython to split our multichain PDB into 3 different PDB files:\n"]},{"cell_type":"code","metadata":{"id":"yZGHA0Jp3Ye_"},"source":["#Here, we import Bio.PDB to use it to manipulate PDB files\n","from Bio.PDB import *\n","#And we set up a parser for our PDB\n","parser = PDBParser()\n","io=PDBIO()\n","structure = parser.get_structure('X', '6ane.pdb')\n","\n","#This will separate each chain into its own PDB file\n","for chain in structure.get_chains():\n","    io.set_structure(chain)\n","    io.save(\"6ane_\" + chain.get_id() + \".pdb\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"kKnC6NyPXez3"},"source":["## Part 2. Visualizing our protein structure using py3Dmol"]},{"cell_type":"markdown","metadata":{"id":"IxwUF4PdhHDM"},"source":["1. To visualize this structure, we will load our PDB file (i.e. 6ane) onto **py3Dmol**, a pyhton implementation of [**3Dmol.js**](http://3dmol.csb.pitt.edu/). You can read the full documentation of 3Dmol.js [here](https://3dmol.csb.pitt.edu/doc/index.html), which is very similar to the one that we are going to use in this tutorial besides some extra stuff here and there.\n","\n","  In the following code cell we take a line by line look to the different commands needed to visualize two of the chains that you extracted from 6ANE. Please note that the text inside each parenthesis follows a JSON data structure, which is reviewed in detail [here](https://towardsdatascience.com/an-introduction-to-json-c9acb464f43e). \n","\n","  Pay special attention to the following **py3Dmol** classes and what they do: _**addmodel**_, _**setBackground**_ and _**setStyle**_.\n","\n","üí° **HINT:** The syntax for loading different structures and properties is really straightforward and self explanatory. Even so, remember to be carefull when opening and closing {} and () and using quotes when needed."]},{"cell_type":"code","metadata":{"id":"0M8qd0sTyb9p"},"source":["#Importing py3Dmol for safety\n","import py3Dmol\n","\n","#First we assign the py3Dmol.view as view\n","view=py3Dmol.view()\n","\n","#The following lines are used to add the addModel class\n","#to read the PDB files of chain B and C\n","view.addModel(open('6ane_B.pdb', 'r').read(),'pdb')\n","view.addModel(open('6ane_C.pdb', 'r').read(),'pdb')\n","\n","#Here we set the background color as white\n","view.setBackgroundColor('white')\n","\n","#Here we set the visualization style for chain B and C\n","view.setStyle({'chain':'B'},{'cartoon': {'color':'purple'}})\n","view.setStyle({'chain':'C'},{'cartoon': {'color':'yellow'}})\n","\n","#Zooming into all visualized structures \n","view.zoomTo()\n","\n","#And we finally visualize the structures using the command below\n","view.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"unujNM2WInSn"},"source":["2. As we can see, we are able to load chains B and C of 6ANE in cartoon representation, but they are far from each other.\n","\n","  If we want to compare the conformations of different residues in these two chains (such as the conformation of the wobbly tryptophan residue mentioned before), we have to **superimpose the structures**. Thus, we are going to use **biopython** again to:\n","\n","  1. Superimpose 6ANE chain B to chain C\n","  2. Print the RMSD\n","  3. Save the superimposed PDB of chain B"]},{"cell_type":"code","metadata":{"id":"GpJplQzgHaQl"},"source":["#The following code was created by Anders Steen Christensen\n","#from the University of Basel and is available at\n","#https://gist.github.com/andersx/6354971\n","\n","import Bio.PDB\n","import os\n","\n","# Select what residue numbers you wish to align\n","# and put them in a list\n","start_id = 1\n","end_id   = 262\n","atoms_to_be_aligned = range(start_id, end_id + 1)\n","\n","# Start the parser\n","pdb_parser = Bio.PDB.PDBParser(QUIET = True)\n","\n","# Get the structures\n","ref_structure = pdb_parser.get_structure(\"reference\", \"6ane_C.pdb\")\n","sample_structure = pdb_parser.get_structure(\"sample\", \"6ane_B.pdb\")\n","\n","# Use the first model in the pdb-files for alignment\n","# Change the number 0 if you want to align to another structure\n","ref_model    = ref_structure[0]\n","sample_model = sample_structure[0]\n","\n","# Make a list of the atoms (in the structures) you wish to align.\n","# In this case we use CA atoms whose index is in the specified range\n","ref_atoms = []\n","sample_atoms = []\n","\n","# Iterate of all chains in the model in order to find all residues\n","for ref_chain in ref_model:\n","  # Iterate of all residues in each model in order to find proper atoms\n","  for ref_res in ref_chain:\n","    # Check if residue number ( .get_id() ) is in the list\n","    if ref_res.get_id()[1] in atoms_to_be_aligned:\n","      # Append CA atom to list\n","      ref_atoms.append(ref_res['CA'])\n","\n","# Do the same for the sample structure\n","for sample_chain in sample_model:\n","  for sample_res in sample_chain:\n","    if sample_res.get_id()[1] in atoms_to_be_aligned:\n","      sample_atoms.append(sample_res['CA'])\n","\n","# Now we initiate the superimposer:\n","super_imposer = Bio.PDB.Superimposer()\n","super_imposer.set_atoms(ref_atoms, sample_atoms)\n","super_imposer.apply(sample_model.get_atoms())\n","\n","# Print RMSD:\n","print('The calculated RMSD is:')\n","print (str(super_imposer.rms) + ' √Ö')\n","\n","# Save the aligned version of one of the chains of 6ANE\n","io = Bio.PDB.PDBIO()\n","io.set_structure(sample_structure) \n","io.save(\"6ane_B_aligned.pdb\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["üí°**HINT**: This code works great for superimposing structures with the same number of residues and for homologous proteins for which the residue range that is best for an alignment is known, but it is not appropriate for structures that have difference in length (meaning that they have indels). When working with homologous proteins, you might need to perform a Sequence-based structural alignment on biopython instead, using a script such as the one found [here](https://gist.github.com/fomightez/2601c0f5a13b85cd21b9377169c79836)."],"metadata":{"id":"7aBcZ0Kju2ju"}},{"cell_type":"markdown","metadata":{"id":"SUr08gPrOd3Z"},"source":["3. Now we can load the superimposed structure into **py3dmol** with the code cell show below. Please try to familiarize yourself with these lines by changing the colors of the visualization at will"]},{"cell_type":"code","metadata":{"id":"jWHfH9RyJR0D"},"source":["#Importing py3Dmol for safety\n","import py3Dmol\n","\n","#First we assign the py3Dmol.view as view\n","view=py3Dmol.view()\n","\n","#The following lines are used to add the addModel class\n","#to read the PDB files of the chain B (now aligned against chain C) and C\n","view.addModel(open('6ane_B_aligned.pdb', 'r').read(),'pdb')\n","view.addModel(open('6ane_C.pdb', 'r').read(),'pdb')\n","\n","#Here we set the background color as white\n","view.setBackgroundColor('white')\n","\n","#Here we set the visualization style for chain B and C\n","view.setStyle({'chain':'B'},{'cartoon': {'color':'purple'}})\n","view.setStyle({'chain':'C'},{'cartoon': {'color':'yellow'}})\n","\n","#Zooming into all visualized structures \n","view.zoomTo()\n","\n","#And we finally visualize the structures using the command below\n","view.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"pHS8q5V7Jbqm"},"source":["4. Now we want to inspect the different conformers of Tryptophan 158 between chains B and C of 6ANE. For this, we can select specific residues using _**addStyle**_ classes as shown below:\n","\n","üí° **HINT:** You can use different **color schemes** for stick representations. They can be used to control the color of carbon atoms and are usually indicated by a \"Carbon\" suffix. For example, _**blueCarbon**_ shows all carbon atoms of a stick representation in blue, _**greenCarbon**_ shows all carbon atoms in green, etc.\n","\n"]},{"cell_type":"code","metadata":{"id":"QORgw17cJ3Ge"},"source":["#Importing py3Dmol for safety\n","import py3Dmol\n","\n","#First we assign the py3Dmol.view as view\n","view=py3Dmol.view()\n","\n","#The following lines are used to add the addModel class\n","#to read the PDB files of the chain B (now aligned against chain C) and C\n","view.addModel(open('6ane_B_aligned.pdb', 'r').read(),'pdb')\n","view.addModel(open('6ane_C.pdb', 'r').read(),'pdb')\n","\n","#Here we set the background color as white\n","view.setBackgroundColor('white')\n","\n","#Set a visualization style for chain B\n","view.setStyle({'chain':'B'},{'cartoon': {'color':'purple'}})\n","\n","#Add a visualization style for residue 158 in chain B\n","view.addStyle({'chain':'B','resi':158},{'stick':{'colorscheme':'grayCarbon'}})\n","\n","#Set a visualization style for chain C\n","view.setStyle({'chain':'C'},{'cartoon': {'color':'yellow'}})\n","\n","#Add a visualization style for residue 158 in chain C\n","view.addStyle({'chain':'C','resi':158},{'stick':{'colorscheme':'skyblueCarbon'}})\n","\n","#Zooming into all visualized structures \n","view.zoomTo()\n","\n","#And we finally visualize the structures using the command below\n","view.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"LA5nb9ZkNzo6"},"source":["5. Sometimes we might want to show several residues in our visualization. Instead of writing a line of code per residue, py3Dmol can also work with residue selections:\n","\n","  * A sequence of residues can be given as a range usign square brackets\n","  \n","    [\"158-168\"]\n","\n","  * The same can be done for non-sequential residues using comma-separated numbers: \n","\n","    [158,168,129,182]\n","\n","  How can we show the catalytic residues of 6ANE, corresponding to residue numbers 133, 179 and 210, and the Tryptophan residue in position 158 with different colors? **Do it yourself** by appropriately setting up these residues and their color scheme representation in the following lines of code"]},{"cell_type":"code","metadata":{"id":"EqnZi3WkN0e1"},"source":["#Importing py3Dmol for safety\n","import py3Dmol\n","\n","#First we assign the py3Dmol.view as view\n","view=py3Dmol.view()\n","\n","#The following lines are used to add the addModel class\n","#to read the PDB files of the chain B (now aligned against chain C) and C\n","view.addModel(open('6ane_B_aligned.pdb', 'r').read(),'pdb')\n","view.addModel(open('6ane_C.pdb', 'r').read(),'pdb')\n","\n","#Here we set the background color as white\n","view.setBackgroundColor('white')\n","\n","#Setting style for chain B\n","view.setStyle({'chain':'B'},\n","              {'cartoon': {'color':'purple'}}\n","             )\n","#Add your catalytic residues and choose a color scheme in the next line\n","view.addStyle({'chain':'B','resi':CHANGE ME},\n","              {'stick':{'colorscheme':CHANGE ME}})\n","\n","#Setting style for chain C\n","view.setStyle({'chain':'C'},\n","              {'cartoon': {'color':'yellow'}}\n","             )\n","#Add a different representation for residue 158 with a different color scheme \n","view.addStyle({'chain':'C','resi':CHANGE ME},\n","              {'stick':{'colorscheme':CHANGE ME}}\n","             )\n","#Zooming into all visualized structures \n","view.zoomTo()\n","\n","#And we finally visualize the structures using the command below\n","view.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"1b88ruujOsp-"},"source":["6. What if I need to select and show residues **around a given distance radius** of Trp158? You can use **_within_** selections in your _addStyle_ classes as shown in the following lines of code"]},{"cell_type":"code","metadata":{"id":"DroiHRaGOxle"},"source":["#Importing py3Dmol for safety\n","import py3Dmol\n","\n","#First we assign the py3Dmol.view as view\n","view=py3Dmol.view()\n","\n","#The following lines are used to add the addModel class\n","#to read the PDB files of the chain B (now aligned against chain C) and C\n","view.addModel(open('6ane_B_aligned.pdb', 'r').read(),'pdb')\n","view.addModel(open('6ane_C.pdb', 'r').read(),'pdb')\n","\n","#Here we set the background color as white\n","view.setBackgroundColor('white')\n","\n","#Setting styles for chains B and C \n","view.setStyle({'chain':'B'},\n","              {'cartoon': {'color':'purple'}})\n","view.setStyle({'chain':'C'},\n","              {'cartoon': {'color':'yellow'}})\n","#See residues that are a distance of 7 A from the residue 158\n","view.addStyle({'within':{'distance': 7,\n","                         'sel':{'resi':158}\n","                         }\n","               }\n","              ,{'stick':{'colorscheme':'greenCarbon'}\n","                }\n","              )\n","#After you made your selection, change colors for the Trp from both chains\n","view.addStyle({'chain':'B','resi':158},\n","              {'stick':{'colorscheme':'blueCarbon'}})\n","view.addStyle({'chain':'C','resi':158},\n","              {'stick':{'colorscheme':'skyblueCarbon'}})\n","\n","#Zooming into all visualized structures \n","view.zoomTo()\n","\n","#And we finally visualize the structures using the command below\n","view.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"ugvHbfvpO_Ja"},"source":["7. What about other representations, such as the **van der Waals** representation of the atoms? You can display the atom surface using vDW through the _**addSurface**_ class, as shown in the following line of code and used in the cell code below.\n","\n","  ```\n","  view.addSurface(py3Dmol.VDW,{'opacity':0.7,'color':'white'}, {'chain':'A'})\n","  ```"]},{"cell_type":"code","metadata":{"id":"XtvaMjtJPNut"},"source":["#Importing py3Dmol for safety\n","import py3Dmol\n","\n","#The first commands should be somewhat trivial at this point\n","view=py3Dmol.view()\n","view.addModel(open('6ane_B_aligned.pdb', 'r').read(),'pdb')\n","view.addModel(open('6ane_C.pdb', 'r').read(),'pdb')\n","view.setBackgroundColor('white')\n","\n","#Cartoon representations\n","view.setStyle({'chain':'B'},{'cartoon': {'color':'blue'}})\n","view.setStyle({'chain':'C'},{'cartoon': {'color':'skyblue'}})\n","\n","#Stick representations\n","view.addStyle({'within':{'distance': 7, 'sel':{'resi':158}}},{'stick':{'colorscheme':'grayCarbon'}})\n","view.addStyle({'chain':'B','resi':158},{'stick':{'colorscheme':'purpleCarbon'}})\n","view.addStyle({'chain':'C','resi':158},{'stick':{'colorscheme':'greenCarbon'}})\n","\n","#VDW Surface\n","view.addSurface(py3Dmol.VDW,{'opacity':0.7,'color':'white'}, {'chain':'B'})\n","view.addSurface(py3Dmol.VDW,{'opacity':0.7,'color':'yellow'}, {'chain':'C'})\n","\n","#Zooming and visualizing\n","view.zoomTo()\n","view.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"KNmqUEhBQLQR"},"source":["8. Finally, if you want to identify the residues in your visualization, you can also add **text labels** using the _**addResLabels**_ class."]},{"cell_type":"code","metadata":{"id":"7-qhLPGuQO6J"},"source":["#Importing py3Dmol for safety\n","import py3Dmol\n","\n","#The first commands should be somewhat trivial at this point\n","view=py3Dmol.view()\n","view.addModel(open('6ane_B_aligned.pdb', 'r').read(),'pdb')\n","view.addModel(open('6ane_C.pdb', 'r').read(),'pdb')\n","view.setBackgroundColor('white')\n","\n","#Cartoon and stick representations\n","view.setStyle({'chain':'B'},{'cartoon': {'color':'purple'}})\n","view.addStyle({'chain':'B','resi':158},{'stick':{'colorscheme':'blueCarbon'}})\n","view.setStyle({'chain':'C'},{'cartoon': {'color':'yellow'}})\n","view.addStyle({'chain':'C','resi':158},{'stick':{'colorscheme':'skyblueCarbon'}})\n","\n","#Adding residue labels \n","view.addResLabels({'chain':'B','resi':158},{'fontOpacity':1,'showBackground':'false','fontColor':'black'})\n","\n","#You can also uncomment the following line to make the model spin\n","#view.spin(\"y\")\n","\n","#Zooming and visualizing\n","view.zoomTo()\n","view.show()\n","\n","#EXPERIMENTAL TRICK!!!\n","#The following trick sets an input and a 10 sec delay to print out a PNG\n","#You can right click on the image to save it\n","\n","#import time\n","#input(\"Press enter. Now you have 10 seconds to choose your visualization\")\n","#view.show()\n","#time.sleep(10)    \n","#view.png()"],"execution_count":null,"outputs":[]}]}