{"nbformat":4,"nbformat_minor":0,"metadata":{"accelerator":"GPU","colab":{"provenance":[{"file_id":"https://github.com/pb3lab/ibm3202/blob/master/tutorials/2022/lab15_CF-GROMACS.ipynb","timestamp":1665426870245}],"collapsed_sections":[],"toc_visible":true},"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.10"}},"cells":[{"cell_type":"markdown","source":["# Day 2 â€“ Tutorial 03:Â Coupling AF2 and GROMACS for protein structure prediction and MD simulations"],"metadata":{"id":"fW2rvNoKwz74"}},{"cell_type":"markdown","source":["### Overview\n","\n","In this hands-on tutorial, we will first employ **ColabFold**, developed by Milot Mirdita, Sergey Ovchinnikov and Martin Steinegger to predict the structure of a given protein on **Google Colab**. ColabFold is, in short, an implementation of **AlphaFold2** using **MMseqs2** and **HHsearch**, enabling ultra fast search and cluster huge protein and nucleotide sequence sets, which is key for accelerating the generation of the multiple sequence alignment (**MSA**) for protein structure prediction using AlphaFold2 on Google Colab.\n","\n","We highly recommend that you check out the [ColabFold GitHub](https://github.com/sokrypton/ColabFold) and read their manuscript for more information:\n","\n","- Mirdita M, SchÃ¼tze K, Moriwaki Y, Heo L, Ovchinnikov S, Steinegger M. ColabFold: Making protein folding accessible to all. [*Nature Methods*, 2022, 19, 6, 679-682](https://www.nature.com/articles/s41592-022-01488-1) \n","\n","Once the protein structure is predicted, we will install **GROMACS**, an MD simulation package, which we will use to set-up and perform MD simulations on our predicted protein. We will visualize our protein structure using **py3Dmol**, while the simulation trajectories from our MD production runs will be visualized in a web version of **NGLview**.\n","\n","For more detailed information on other analysis of the motions of our protein, we highly recommend to check the [GROMACS tutorials](http://www.mdtutorials.com/gmx/) developed by Justin Lemkul and our implementation of some of these analyses on Google Colab in our [Molecular Modeling and Simulation GitHub](https://github.com/pb3lab/ibm3202), as well as our article:\n","\n","- Engelberger F, Galaz-Davison P, Bravo G, Rivera M, Ramirez-Sarmiento CA. Developing and Implementing Cloud-Based Tutorials That Combine Bioinformatics Software, Interactive Coding, and Visualization Exercises for Distance Learning on Structural Bioinformatics. [*J Chem Educ*, 2021, 98, 5, 1801â€“1807](https://pubs.acs.org/doi/full/10.1021/acs.jchemed.1c00022)\n"],"metadata":{"id":"Ht2PzspdxKs3"}},{"cell_type":"markdown","metadata":{"id":"UGUBLzB3C6WN","pycharm":{"name":"#%% md\n"}},"source":["## Part I. Protein Structure Prediction on ColabFold\n","\n","We will run Colab on a readily defined sequence using the default parameters (except for the recycling, which is now set to 1). For continuity with the tutorials, we will model the structure of a potential ancestral sequence in the node where the human FoxP and FoxO proteins split.\n","\n","ðŸ˜±**EMERGENCY BACKUP**: \"But I did not save that sequence!\". Do not worry, we got you:\n","\n","\n","```\n",">AncOP\n","RPPFSYASLITQAIEESPEKRLTLNEIYNWIMRNFPYFRDKGDSNNAAGWKNSIRHNLSLHKCFVKVPREHDESTGKGSFWTID\n","```\n","\n","Therefore, the instructions for this tutorial are:\n","\n","1. Paste your protein sequence in the input field.\n","2. Run ColabFold by clicking on the \"Play\" button on the left of each cell.\n","3. The pipeline consists of 5 steps. The currently running step is indicated by a circle with a stop sign next to it.\n","\n","Once the final protein structure is predicted, the results will be visually presented on ColabFold and a ZIP file will be generated, which contains the following information:\n","\n","1. PDB formatted structures sorted by avg. pLDDT and complexes are sorted by pTMscore (unrelaxed and relaxed if `use_amber` is enabled).\n","2. Plots of the model quality.\n","3. Plots of the MSA coverage.\n","4. Parameter log file.\n","5. A3M formatted input MSA.\n","6. A `predicted_aligned_error_v1.json` using [AlphaFold-DB's format](https://alphafold.ebi.ac.uk/faq#faq-7) and a `scores.json` for each model which contains an array (list of lists) for PAE, a list with the average pLDDT and the pTMscore.\n","7. BibTeX file with citations for all used tools and databases.\n","\n","At the end of the job a download modal box will pop up with a `jobname.result.zip` file. Additionally, if the `save_to_google_drive` option was selected, the `jobname.result.zip` will be uploaded to your Google Drive.\n","\n","For the description of additional options (prediction of protein complexes, custom MSAs, custom templates, etc) please check out the latest version of [ColabFold](http://colabfold.com).\n"]},{"cell_type":"code","metadata":{"id":"kOblAo-xetgx","cellView":"form"},"source":["#@title 1. Input protein sequence(s), then hit `Runtime` -> `Run all`\n","from google.colab import files\n","import os.path\n","import re\n","import hashlib\n","import random\n","\n","def add_hash(x,y):\n","  return x+\"_\"+hashlib.sha1(y.encode()).hexdigest()[:5]\n","\n","query_sequence = 'THISISNOTAPROTEINSEQUENCE' #@param {type:\"string\"}\n","#@markdown  - Use `:` to specify inter-protein chainbreaks for **modeling complexes** (supports homo- and hetro-oligomers). For example **PI...SK:PI...SK** for a homodimer\n","\n","# remove whitespaces\n","query_sequence = \"\".join(query_sequence.split())\n","\n","jobname = 'test' #@param {type:\"string\"}\n","# remove whitespaces\n","basejobname = \"\".join(jobname.split())\n","basejobname = re.sub(r'\\W+', '', basejobname)\n","jobname = add_hash(basejobname, query_sequence)\n","while os.path.isfile(f\"{jobname}.csv\"):\n","  jobname = add_hash(basejobname, ''.join(random.sample(query_sequence,len(query_sequence))))\n","\n","with open(f\"{jobname}.csv\", \"w\") as text_file:\n","    text_file.write(f\"id,sequence\\n{jobname},{query_sequence}\")\n","\n","queries_path=f\"{jobname}.csv\"\n","\n","# number of models to use\n","use_amber = False #@param {type:\"boolean\"}\n","template_mode = \"none\" #@param [\"none\", \"pdb70\",\"custom\"]\n","#@markdown - \"none\" = no template information is used, \"pdb70\" = detect templates in pdb70, \"custom\" - upload and search own templates (PDB or mmCIF format, see [notes below](#custom_templates))\n","\n","if template_mode == \"pdb70\":\n","  use_templates = True\n","  custom_template_path = None\n","elif template_mode == \"custom\":\n","  custom_template_path = f\"{jobname}_template\"\n","  os.mkdir(custom_template_path)\n","  uploaded = files.upload()\n","  use_templates = True\n","  for fn in uploaded.keys():\n","    os.rename(fn, f\"{jobname}_template/{fn}\")\n","else:\n","  custom_template_path = None\n","  use_templates = False\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@markdown ### 2. MSA options (custom MSA upload, single sequence, pairing mode)\n","msa_mode = \"MMseqs2 (UniRef+Environmental)\" #@param [\"MMseqs2 (UniRef+Environmental)\", \"MMseqs2 (UniRef only)\",\"single_sequence\",\"custom\"]\n","pair_mode = \"unpaired+paired\" #@param [\"unpaired+paired\",\"paired\",\"unpaired\"] {type:\"string\"}\n","#@markdown - \"unpaired+paired\" = pair sequences from same species + unpaired MSA, \"unpaired\" = seperate MSA for each chain, \"paired\" - only use paired sequences.\n","\n","# decide which a3m to use\n","if msa_mode.startswith(\"MMseqs2\"):\n","  a3m_file = f\"{jobname}.a3m\"\n","elif msa_mode == \"custom\":\n","  a3m_file = f\"{jobname}.custom.a3m\"\n","  if not os.path.isfile(a3m_file):\n","    custom_msa_dict = files.upload()\n","    custom_msa = list(custom_msa_dict.keys())[0]\n","    header = 0\n","    import fileinput\n","    for line in fileinput.FileInput(custom_msa,inplace=1):\n","      if line.startswith(\">\"):\n","         header = header + 1\n","      if not line.rstrip():\n","        continue\n","      if line.startswith(\">\") == False and header == 1:\n","         query_sequence = line.rstrip()\n","      print(line, end='')\n","\n","    os.rename(custom_msa, a3m_file)\n","    queries_path=a3m_file\n","    print(f\"moving {custom_msa} to {a3m_file}\")\n","else:\n","  a3m_file = f\"{jobname}.single_sequence.a3m\"\n","  with open(a3m_file, \"w\") as text_file:\n","    text_file.write(\">1\\n%s\" % query_sequence)"],"metadata":{"cellView":"form","id":"C2_sh2uAonJH"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@markdown ### 3. Advanced settings\n","model_type = \"auto\" #@param [\"auto\", \"AlphaFold2-ptm\", \"AlphaFold2-multimer-v1\", \"AlphaFold2-multimer-v2\"]\n","#@markdown - \"auto\" = protein structure prediction using \"AlphaFold2-ptm\" and complex prediction \"AlphaFold-multimer-v2\". For complexes \"AlphaFold-multimer-v[1,2]\" and \"AlphaFold-ptm\" can be used.\n","num_recycles = 1 #@param [1,3,6,12,24,48] {type:\"raw\"}\n","save_to_google_drive = False #@param {type:\"boolean\"}\n","\n","#@markdown -  if the save_to_google_drive option was selected, the result zip will be uploaded to your Google Drive\n","dpi = 200 #@param {type:\"integer\"}\n","#@markdown - set dpi for image resolution\n","\n","#@markdown Don't forget to hit `Runtime` -> `Run all` after updating the form.\n","\n","\n","if save_to_google_drive:\n","  from pydrive.drive import GoogleDrive\n","  from pydrive.auth import GoogleAuth\n","  from google.colab import auth\n","  from oauth2client.client import GoogleCredentials\n","  auth.authenticate_user()\n","  gauth = GoogleAuth()\n","  gauth.credentials = GoogleCredentials.get_application_default()\n","  drive = GoogleDrive(gauth)\n","  print(\"You are logged into Google Drive and are good to go!\")"],"metadata":{"cellView":"form","id":"ADDuaolKmjGW"},"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"iccGdbe_Pmt9","pycharm":{"name":"#%%\n"},"cellView":"form"},"source":["#@title 4. Install AF2 dependencies\n","%%bash -s $use_amber $use_templates\n","\n","set -e\n","\n","USE_AMBER=$1\n","USE_TEMPLATES=$2\n","\n","if [ ! -f COLABFOLD_READY ]; then\n","  # install dependencies\n","  # We have to use \"--no-warn-conflicts\" because colab already has a lot preinstalled with requirements different to ours\n","  pip install -q --no-warn-conflicts \"colabfold[alphafold-minus-jax] @ git+https://github.com/sokrypton/ColabFold\"\n","  # high risk high gain\n","  pip install -q \"jax[cuda11_cudnn805]>=0.3.8,<0.4\" -f https://storage.googleapis.com/jax-releases/jax_releases.html\n","  touch COLABFOLD_READY\n","fi\n","\n","# setup conda\n","if [ ${USE_AMBER} == \"True\" ] || [ ${USE_TEMPLATES} == \"True\" ]; then\n","  if [ ! -f CONDA_READY ]; then\n","    wget -qnc https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh\n","    bash Miniconda3-latest-Linux-x86_64.sh -bfp /usr/local 2>&1 1>/dev/null\n","    rm Miniconda3-latest-Linux-x86_64.sh\n","    touch CONDA_READY\n","  fi\n","fi\n","# setup template search\n","if [ ${USE_TEMPLATES} == \"True\" ] && [ ! -f HH_READY ]; then\n","  conda install -y -q -c conda-forge -c bioconda kalign2=2.04 hhsuite=3.3.0 python=3.7 2>&1 1>/dev/null\n","  touch HH_READY\n","fi\n","# setup openmm for amber refinement\n","if [ ${USE_AMBER} == \"True\" ] && [ ! -f AMBER_READY ]; then\n","  conda install -y -q -c conda-forge openmm=7.5.1 python=3.7 pdbfixer 2>&1 1>/dev/null\n","  touch AMBER_READY\n","fi"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"_sztQyz29DIC","cellView":"form"},"source":["#@title 5. Run Prediction\n","\n","import sys\n","\n","from colabfold.download import download_alphafold_params, default_data_dir\n","from colabfold.utils import setup_logging\n","from colabfold.batch import get_queries, run, set_model_type\n","K80_chk = !nvidia-smi | grep \"Tesla K80\" | wc -l\n","if \"1\" in K80_chk:\n","  print(\"WARNING: found GPU Tesla K80: limited to total length < 1000\")\n","  if \"TF_FORCE_UNIFIED_MEMORY\" in os.environ:\n","    del os.environ[\"TF_FORCE_UNIFIED_MEMORY\"]\n","  if \"XLA_PYTHON_CLIENT_MEM_FRACTION\" in os.environ:\n","    del os.environ[\"XLA_PYTHON_CLIENT_MEM_FRACTION\"]\n","\n","from colabfold.colabfold import plot_protein\n","from pathlib import Path\n","import matplotlib.pyplot as plt\n","\n","\n","# For some reason we need that to get pdbfixer to import\n","if use_amber and '/usr/local/lib/python3.7/site-packages/' not in sys.path:\n","    sys.path.insert(0, '/usr/local/lib/python3.7/site-packages/')\n","\n","def prediction_callback(unrelaxed_protein, length, prediction_result, input_features, type):\n","  fig = plot_protein(unrelaxed_protein, Ls=length, dpi=150)\n","  plt.show()\n","  plt.close()\n","\n","result_dir=\".\"\n","setup_logging(Path(\".\").joinpath(\"log.txt\"))\n","queries, is_complex = get_queries(queries_path)\n","model_type = set_model_type(is_complex, model_type)\n","download_alphafold_params(model_type, Path(\".\"))\n","run(\n","    queries=queries,\n","    result_dir=result_dir,\n","    use_templates=use_templates,\n","    custom_template_path=custom_template_path,\n","    use_amber=use_amber,\n","    msa_mode=msa_mode,    \n","    model_type=model_type,\n","    num_models=5,\n","    num_recycles=num_recycles,\n","    model_order=[1, 2, 3, 4, 5],\n","    is_complex=is_complex,\n","    data_dir=Path(\".\"),\n","    keep_existing_results=False,\n","    recompile_padding=1.0,\n","    rank_by=\"auto\",\n","    pair_mode=pair_mode,\n","    stop_at_score=float(100),\n","    prediction_callback=prediction_callback,\n","    dpi=dpi\n",")"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"KK7X9T44pWb7","cellView":"form"},"source":["#@title 6. Display 3D structure {run: \"auto\"}\n","import py3Dmol\n","import glob\n","import matplotlib.pyplot as plt\n","from colabfold.colabfold import plot_plddt_legend\n","rank_num = 1 #@param [\"1\", \"2\", \"3\", \"4\", \"5\"] {type:\"raw\"}\n","color = \"lDDT\" #@param [\"chain\", \"lDDT\", \"rainbow\"]\n","show_sidechains = False #@param {type:\"boolean\"}\n","show_mainchains = False #@param {type:\"boolean\"}\n","\n","jobname_prefix = \".custom\" if msa_mode == \"custom\" else \"\"\n","if use_amber:\n","  pdb_filename = f\"{jobname}{jobname_prefix}_relaxed_rank_{rank_num}_model_*.pdb\"\n","else:\n","  pdb_filename = f\"{jobname}{jobname_prefix}_unrelaxed_rank_{rank_num}_model_*.pdb\"\n","\n","pdb_file = glob.glob(pdb_filename)\n","\n","def show_pdb(rank_num=1, show_sidechains=False, show_mainchains=False, color=\"lDDT\"):\n","  model_name = f\"rank_{rank_num}\"\n","  view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js',)\n","  view.addModel(open(pdb_file[0],'r').read(),'pdb')\n","\n","  if color == \"lDDT\":\n","    view.setStyle({'cartoon': {'colorscheme': {'prop':'b','gradient': 'roygb','min':50,'max':90}}})\n","  elif color == \"rainbow\":\n","    view.setStyle({'cartoon': {'color':'spectrum'}})\n","  elif color == \"chain\":\n","    chains = len(queries[0][1]) + 1 if is_complex else 1\n","    for n,chain,color in zip(range(chains),list(\"ABCDEFGH\"),\n","                     [\"lime\",\"cyan\",\"magenta\",\"yellow\",\"salmon\",\"white\",\"blue\",\"orange\"]):\n","      view.setStyle({'chain':chain},{'cartoon': {'color':color}})\n","  if show_sidechains:\n","    BB = ['C','O','N']\n","    view.addStyle({'and':[{'resn':[\"GLY\",\"PRO\"],'invert':True},{'atom':BB,'invert':True}]},\n","                        {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n","    view.addStyle({'and':[{'resn':\"GLY\"},{'atom':'CA'}]},\n","                        {'sphere':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n","    view.addStyle({'and':[{'resn':\"PRO\"},{'atom':['C','O'],'invert':True}]},\n","                        {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})  \n","  if show_mainchains:\n","    BB = ['C','O','N','CA']\n","    view.addStyle({'atom':BB},{'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n","\n","  view.zoomTo()\n","  return view\n","\n","\n","show_pdb(rank_num,show_sidechains, show_mainchains, color).show()\n","if color == \"lDDT\":\n","  plot_plddt_legend().show() "],"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@title 7. Generate plots {run: \"auto\"}\n","from IPython.display import display, HTML\n","import base64\n","from html import escape\n","\n","# see: https://stackoverflow.com/a/53688522\n","def image_to_data_url(filename):\n","  ext = filename.split('.')[-1]\n","  prefix = f'data:image/{ext};base64,'\n","  with open(filename, 'rb') as f:\n","    img = f.read()\n","  return prefix + base64.b64encode(img).decode('utf-8')\n","\n","pae = image_to_data_url(f\"{jobname}{jobname_prefix}_PAE.png\")\n","cov = image_to_data_url(f\"{jobname}{jobname_prefix}_coverage.png\")\n","plddt = image_to_data_url(f\"{jobname}{jobname_prefix}_plddt.png\")\n","display(HTML(f\"\"\"\n","<style>\n","  img {{\n","    float:left;\n","  }}\n","  .full {{\n","    max-width:100%;\n","  }}\n","  .half {{\n","    max-width:50%;\n","  }}\n","  @media (max-width:640px) {{\n","    .half {{\n","      max-width:100%;\n","    }}\n","  }}\n","</style>\n","<div style=\"max-width:90%; padding:2em;\">\n","  <h1>Plots for {escape(jobname)}</h1>\n","  <img src=\"{pae}\" class=\"full\" />\n","  <img src=\"{cov}\" class=\"half\" />\n","  <img src=\"{plddt}\" class=\"half\" />\n","</div>\n","\"\"\"))\n"],"metadata":{"cellView":"form","id":"BKfJq1IN271K"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["### Description of the plots\n","*   **Number of sequences per position** - We want to see at least 30 sequences per position,  and for best performance, ideally 100 sequences or more.\n","*   **Predicted lDDT per position** - Model confidence (out of 100) at each position. The higher the better.\n","*   **Predicted Alignment Error** - For homooligomers, this could be a useful metric to assess how confident the model is about the interface. The lower the better."],"metadata":{"id":"nw2Bl2HI2lYe"}},{"cell_type":"code","source":["#@title 8. Package and download results (optional)\n","#@markdown If you are having issues downloading the result archive, try disabling your adblocker and run this cell again. If that fails click on the little folder icon to the left, navigate to file: `jobname.result.zip`, right-click and select \\\"Download\\\" (see [screenshot](https://pbs.twimg.com/media/E6wRW2lWUAEOuoe?format=jpg&name=small)).\n","\n","if msa_mode == \"custom\":\n","  print(\"Don't forget to cite your custom MSA generation method.\")\n","\n","!zip -FSr $jobname\".result.zip\" config.json $jobname*\".json\" $jobname*\".a3m\" $jobname*\"relaxed_rank_\"*\".pdb\" \"cite.bibtex\" $jobname*\".png\"\n","files.download(f\"{jobname}.result.zip\")\n","\n","if save_to_google_drive == True and drive:\n","  uploaded = drive.CreateFile({'title': f\"{jobname}.result.zip\"})\n","  uploaded.SetContentFile(f\"{jobname}.result.zip\")\n","  uploaded.Upload()\n","  print(f\"Uploaded {jobname}.result.zip to Google Drive with ID {uploaded.get('id')}\")"],"metadata":{"cellView":"form","id":"ie8vpyYt3Dno"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Part II. Download and install GROMACS"],"metadata":{"id":"RFVqQRwGDiSB"}},{"cell_type":"markdown","source":["We will first start by setting up **GROMACS** on Google Colab, based on a  previously compiled and installed GROMACS.\n"],"metadata":{"id":"EmRg2spf6Bra"}},{"cell_type":"code","source":["#@title ### Installing GROMACS 2020.6 version\n","!wget https://raw.githubusercontent.com/pb3lab/ibm3202/master/software/gromacs.tar.gz\n","!tar xzf gromacs.tar.gz\n","# It is recommended (and required for GROMACS 2021) to upgrade cmake\n","!pip install cmake --upgrade"],"metadata":{"id":"vmBJRXQJDh9y","cellView":"form"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@title ### Checking that GROMACS 2020.6 runs on Google Colab\n","\n","%%bash\n","source /content/gromacs/bin/GMXRC\n","gmx -h"],"metadata":{"id":"y976RXGIET86","cellView":"form"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Part III. Preparing our MD simulation system\n"],"metadata":{"id":"CNOQQVG-Gc5M"}},{"cell_type":"markdown","source":["### Part III.A. Parameterizing the atoms building up our system for MD simulations"],"metadata":{"id":"Z06yiZio8R-o"}},{"cell_type":"markdown","source":["Now, we will work with GROMACS to parameterize our protein, generating:\n","\n","- A .gro or .pdb coordinate file that contains all the atom types as defined by a given force field (including hydrogens).\n","- A .top topology file containing the parameters for bonds, angles, dihedrals and non-bonded interactions defined by a given force field (potential energy function) to employ in our simulations.\n","\n","We will parameterize our protein using the AMBER99SB-ILDN force field on GROMACS and obtain these files using gmx as shown in the code cell below. This force field is extensively used in MD simulations and has parameters that well-represent the dynamics and flexibility of folded proteins. Notice that the dynamics of highly motile proteins or intrinsically disordered regions is not the main dataset for which this force field was parameterized, and other options may better suit such goals."],"metadata":{"id":"7vS5q8NGGjlr"}},{"cell_type":"code","source":["#@title #### 1. Making a copy of the ColabFold predictions for safety\n","!mkdir /content/results\n","!cp /content/*pdb /content/results/.\n","!cp /content/*json /content/results/.\n","!cp /content/*png /content/results/.\n","!cp /content/results/*rank_1*.pdb /content/structure.pdb"],"metadata":{"id":"yzWtdL4_F3uS","cellView":"form"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@title #### 2. Parameterizing our protein using the AMBER99SB-ILDN force field\n","%%bash\n","source /content/gromacs/bin/GMXRC\n","\n","#Using pdb2gmx to parameterize our PDB with the AMBER forcefield and SPC/E water\n","gmx pdb2gmx -f structure.pdb -o processed.pdb -water spce -ignh -ff amber99sb-ildn -quiet"],"metadata":{"id":"B4q8pnQRGoy7","cellView":"form"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["### Part III.B. Solvating our protein"],"metadata":{"id":"5nQt-KPpHQyr"}},{"cell_type":"markdown","source":["We will now define a periodic box for our simulation system, in which our protein will be centered ,and then fill this box with water molecules, thus solvating our protein. Typically, a **padding distance** of 1.0-1.5 nm around the protein is recommended for globular proteins."],"metadata":{"id":"lWG7tgauHXj7"}},{"cell_type":"code","source":["#@title #### 1. Setting up a cubic box of 1.0 nm padding\n","\n","%%bash\n","source /content/gromacs/bin/GMXRC\n","\n","#Using editconf to create a cubic box with 1.0 nm padding for our solvated system\n","gmx editconf -f processed.pdb -o newbox.pdb -c -d 1.0 -bt cubic -quiet"],"metadata":{"id":"7svfDHR9HeiU","cellView":"form"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@title #### 2. Filling up our cubic box with water molecules\n","%%bash\n","source /content/gromacs/bin/GMXRC\n","\n","#Using solvate to fill up our box with water molecules\n","gmx solvate -cp newbox.pdb -o solv.pdb -p topol.top -quiet"],"metadata":{"id":"TE7cg7GiHxbs","cellView":"form"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"2H5OJEC7EC78"},"source":["Please note that, given the addition of water molecules to our simulation system, we are generating **a new topology file** and **a new coordinate file with added water molecules**"]},{"cell_type":"code","metadata":{"id":"WRB_ek2VTnlP","cellView":"form"},"source":["#@title #### 3. Visualization of our solvated system\n","\n","#First we assign the py3Dmol.view as view\n","view=py3Dmol.view()\n","#The following lines are used to add the addModel class\n","#to read the PDB files\n","view.addModel(open('solv.pdb', 'r').read(),'pdb')\n","#Here we set the background color as white\n","view.setBackgroundColor('white')\n","#Here we set the visualization style and color\n","view.setStyle({'cartoon': {'color':'green'}})\n","#Here we add a style for showing the oxygen from water molecules\n","view.addStyle({'atom':'OW'},{'sphere':{'radius':'0.2'}})\n","#Centering the view on all visible atoms\n","view.zoomTo()\n","#And we finally visualize the structures using the command below\n","view.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"OJkvYoHEPaXA"},"source":["### Part III.C. Adding counterions to neutralize the global charge of the system"]},{"cell_type":"markdown","source":["Now we have a solvated box, but our system has a non-zero charge. We need to neutralize the charges of our simulation system.\n"],"metadata":{"id":"cjWoc1BFTTge"}},{"cell_type":"code","source":["#@title #### 1. The charge of your solvated system is:\n","!grep \"qtot\" topol.top | awk 'END{print $(NF)'}\n"],"metadata":{"cellView":"form","id":"xJL4idUI_pWu"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"IDdhklyOQ4A4"},"source":["Therefore, we will replace water molecules with the counterions required to get the absolute charge of the system to **zero**."]},{"cell_type":"code","metadata":{"id":"UzCySYRYRpF1","cellView":"form"},"source":["#@title #### 2. Neutralizing the simulation system\n","%%bash\n","wget https://raw.githubusercontent.com/pb3lab/ibm3202/master/files/ions.mdp\n","source /content/gromacs/bin/GMXRC\n","\n","#Using grompp and an MD instruction file to add counterions to our system\n","gmx grompp -f ions.mdp -c solv.pdb -p topol.top -o ions.tpr -quiet\n","\n","#This is a trick to provide interactive options to gmx\n","echo \"SOL\" > options\n","echo \" \" >> options\n","\n","#Using genion and the tpr to add counterions to our solvated system\n","gmx genion -s ions.tpr -o solv_ions.pdb -p topol.top -pname NA -nname CL -neutral < options"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"c6pJ51svIwvM"},"source":["## Part IV. Running an MD simulation"]},{"cell_type":"markdown","source":["### Part IV.A. Minimization and equilibration of our system"],"metadata":{"id":"dxY7BAO1tdkC"}},{"cell_type":"markdown","metadata":{"id":"2pau9_KLb8pY"},"source":["Now, we are ready to perform the minimization of our system to eliminate the high energy and forces due to bad initial coordinates, and its equilibration at constant pressure and temperature (NPT ensemble).\n","\n","In very simple words, each simulation step in GROMACS requires:\n","1. Generating an **MD parameter (.mdp) file** that provides the details of the simulation\n","2. Pre-processing this file using **`grompp`**\n","3. Running the simulation using **`mdrun`**\n","\n","As an example, we will start by downloading an MD instruction file that contains all of the parameters required for the minimization of our system and also print its contents for inspection."]},{"cell_type":"code","metadata":{"id":"fpc6RZfkd2i5","cellView":"form"},"source":["#@title ###1. Download and analyze a GROMACS .mdp file\n","!wget https://raw.githubusercontent.com/pb3lab/ibm3202/master/files/em.mdp\n","!paste em.mdp"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Ik60hrRCJ8gR"},"source":["As you can see, the minimization process involves the use of a **steepest descent** (`steep`) integrator. This integrator, alongside conjugate gradient (`cg`) and a quasi-Newtonian method (`l-bfgs`), are **minimization algorithms** that instead of solving the positions for changes in the gradient (Newtonâ€™s equation of motion), look for changes in position that would **minimize the potential energy**.\n"]},{"cell_type":"code","metadata":{"id":"P2usOEZ4J8Hh","cellView":"form"},"source":["#@title ### 2. Run the energy minimization\n","%%bash\n","source /content/gromacs/bin/GMXRC\n","\n","#Using grompp to prepare our minimization MD\n","gmx grompp -f em.mdp -c solv_ions.pdb -p topol.top -o em.tpr -quiet\n","\n","#Run our minimization\n","gmx mdrun -deffnm em -nb gpu -quiet"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"OzZMS778M9tn"},"source":["Once our minimization is finished, we can check how the potential energy of the system changes over each minimization step."]},{"cell_type":"code","metadata":{"id":"6B29pBeTM8rY","cellView":"form"},"source":["#@title ### 3. Extract the change in potential energy during minimization\n","%%bash\n","source /content/gromacs/bin/GMXRC\n","\n","#This is a trick to provide interactive options to gmx\n","echo \"Potential\" > options\n","echo \" \" >> options\n","\n","#Using energy to extract the potential energy of the system\n","gmx energy -f em.edr -o em_potential.xvg -xvg none -quiet < options"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"4SdH8HvkN9Yp","cellView":"form"},"source":["#@title ### 4. Plot the change in potential energy of the system\n","import matplotlib.pyplot as plt\n","plt.style.use('seaborn-whitegrid')\n","import numpy as np\n","\n","#Reading the text file containing this information\n","data = np.loadtxt('em_potential.xvg')\n","\n","plt.title('Potential Energy during Minimization')\n","plt.xlabel('Energy Minimization Step')\n","plt.ylabel(r'E$_P$ [kJâ€¢mol$^{-1}]$')\n","plt.plot(data[:,0], data[:,1], linestyle='solid', linewidth='2', color='red') \n","plt.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"SDRRj9gGQ2AC"},"source":["Next, we will **equilibrate the energy and density of our system at constant temperature and pressure** before the MD production runs.\n","\n","First, we will equilibrate our system at a target temperature (in our case, 300K) using a thermal bath. The initial velocities for the atoms of our system at our target temperature are obtained through a Maxwell distribution."]},{"cell_type":"code","metadata":{"id":"6SFZAfVUfYAl","cellView":"form"},"source":["#@title ### 5. Downloading an MD parameter file and running an NVT equilibration\n","%%time\n","%%bash\n","wget https://raw.githubusercontent.com/pb3lab/ibm3202/master/files/nvt.mdp\n","\n","#Using grompp to prepare our NVT equilibration MD\n","source /content/gromacs/bin/GMXRC\n","gmx grompp -f nvt.mdp -c em.gro -r em.gro -p topol.top -o nvt.tpr -quiet\n","\n","#Run our NVT equilibration MD\n","source /content/gromacs/bin/GMXRC\n","gmx mdrun -deffnm nvt -nb gpu -quiet"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"bctYB0kwU_o3"},"source":["What we just did corresponds to a simulation setup in which the number of atoms, the volume and the temperature are kept constant: **NVT ensemble**. Thus, the temperature of the system should oscillate around the desired temperature (in our case, 300 K). We will confirm this condition below."]},{"cell_type":"code","metadata":{"id":"onmMm5CJVb6x","cellView":"form"},"source":["#@title ### 6. Extract the change in temperature during the NVT equilibration\n","%%bash\n","source /content/gromacs/bin/GMXRC\n","\n","#This is a trick to provide interactive options to gmx\n","echo \"Temperature\" > options\n","echo \" \" >> options\n","\n","#Using energy to extract the temperature of the system during the NVT equil MD\n","gmx energy -f nvt.edr -o nvt_temp.xvg -xvg none -quiet < options"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"Gpn9-IoUVb65","cellView":"form"},"source":["#@title ### 7. Plotting the change in temperature of the system during the NVT equilibration\n","import matplotlib.pyplot as plt\n","plt.style.use('seaborn-whitegrid')\n","import numpy as np\n","\n","#Reading the text file containing this information\n","data = np.loadtxt('nvt_temp.xvg')\n","\n","plt.title('Temperature during 0.1 ns Equilibration (NVT)')\n","plt.xlabel('Time (ps)')\n","plt.ylabel('Temperature [K]')\n","plt.plot(data[:,0], data[:,1], linestyle='solid', linewidth='2', color='red') \n","plt.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"uzGjrmfZWIxz"},"source":["Lastly, we will equilibrate our system at **constant pressure**, which will maintain the density of our solvent constant so that it matches what we would expect for a protein in solution at atmospheric conditions. Thus, in this case we will be using an ensemble in which the number of atoms, the pressure and the temperature of the system remain constant: **the NPT ensemble**."]},{"cell_type":"code","metadata":{"id":"L5XBoc3o3Iwd","cellView":"form"},"source":["#@title ### 8. Downloading an MD parameter file and running an NPT equilibration \n","%%time\n","%%bash\n","wget https://raw.githubusercontent.com/pb3lab/ibm3202/master/files/npt.mdp\n","\n","#Using grompp to prepare our NPT equilibration MD\n","source /content/gromacs/bin/GMXRC\n","gmx grompp -f npt.mdp -c nvt.gro -r nvt.gro -t nvt.cpt -p topol.top -o npt.tpr -quiet\n","\n","#Run our NPT equilibration MD\n","source /content/gromacs/bin/GMXRC\n","gmx mdrun -deffnm npt -nb gpu -quiet"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"vuu5jtpYW3LH"},"source":["7. Given that we are using an NPT ensemble to maintain our simulation at constant pressure and density, we should check if this is achieved."]},{"cell_type":"code","metadata":{"id":"sLsoQy0fXMCj","cellView":"form"},"source":["#@title ### 9. Extract the change in pressure and density during the NPT equilibration \n","%%bash\n","source /content/gromacs/bin/GMXRC\n","\n","#This is a trick to provide interactive options to gmx\n","echo \"Pressure\" > options\n","echo \"Density\" >> options\n","echo \" \"\n","\n","#Using energy to extract the pressure and density of the system during the NPT equil MD\n","gmx energy -f npt.edr -o npt_press_dens.xvg -xvg none -quiet < options"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"aR-HxTtUXMCp","cellView":"form"},"source":["#@title ### 10. Plotting the change in pressure and density of the system during the NPT equilibration\n","import matplotlib.pyplot as plt\n","plt.style.use('seaborn-whitegrid')\n","import numpy as np\n","\n","#Reading the text files containing this information\n","data1 = np.loadtxt('npt_press_dens.xvg')\n","data2 = np.loadtxt('npt_press_dens.xvg')\n","\n","fig, (ax1, ax2) = plt.subplots(2)\n","fig.suptitle('Pressure and density during 0.1 ns Equilibration (NPT)')\n","\n","ax1.set(ylabel='Pressure [bar]')\n","#ax1.ylim(-250,250)\n","\n","ax2.set(xlabel='Time (ps)', ylabel='Density [kgâ€¢m$^{-3}$]')\n","#ax2.ylim(1000,1020)\n","\n","#Smoothing using Savitzky-Golay\n","from scipy.signal import savgol_filter\n","yhat = savgol_filter(data1[:,1], 21, 5)\n","\n","#Plot raw data and spline interpolation\n","ax1.plot(data1[:,0], data1[:,1], linestyle='solid', linewidth='2', color='red')\n","ax1.plot(data1[:,0], yhat, linestyle='solid', linewidth='2', color='blue') \n","ax2.plot(data2[:,0], data2[:,2], linestyle='solid', linewidth='2', color='red') \n","\n","plt.show()"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"zJaUO1kDdSTG"},"source":["### Part IV.B. Running an MD simulation"]},{"cell_type":"markdown","metadata":{"id":"OA-LlTAYdt1z"},"source":["Now we are ready for running our MD simulations. Due to time constraints and the size of the system, we are only generating **0.1 ns of production runs**, whilst you will find that simulations in current articles often correspond to hundreds of ns."]},{"cell_type":"code","metadata":{"id":"r2K3pzyn3q7G","cellView":"form"},"source":["#@title ### 1. Downloading an MD parameter file and runninng a production MD simulation\n","%%time\n","%%bash\n","wget https://raw.githubusercontent.com/pb3lab/ibm3202/master/files/md.mdp\n","\n","#Using grompp to prepare our production MD\n","source /content/gromacs/bin/GMXRC\n","gmx grompp -f md.mdp -c npt.gro -t npt.cpt -p topol.top -o md_1.tpr -quiet\n","\n","#Run our production MD\n","source /content/gromacs/bin/GMXRC\n","gmx mdrun -deffnm md_1 -nb gpu -quiet"],"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@title ### 2. Analyzing the global (RMSD) and local (RMSF) dynamics of the protein\n","%%bash\n","source /content/gromacs/bin/GMXRC\n","\n","#Commands for RMSD\n","echo \"C-alpha\" > options\n","echo \" \" >> options\n","echo \"C-alpha\" >> options\n","echo \" \" >> options\n","#RMSD calculation\n","gmx rms -s em.gro -f md_1.xtc -xvg none < options\n","#RMSF calculation\n","gmx rmsf -s em.gro -f md_1.xtc -xvg none -res < options"],"metadata":{"cellView":"form","id":"rD_Q9S2McZR4"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@title ### 3. Plotting the RMSD of our short simulation\n","import matplotlib.pyplot as plt\n","plt.style.use('seaborn-whitegrid')\n","import numpy as np\n","\n","#Reading the text file containing this information\n","data = np.loadtxt('rmsd.xvg')\n","\n","plt.title('Ca-RMSD during 0.1 ns production MD')\n","plt.xlabel('Time (ps)')\n","plt.ylabel('Ca-RMSD (nm)')\n","plt.plot(data[:,0], data[:,1], linestyle='solid', linewidth='2', color='red') \n","plt.show()"],"metadata":{"cellView":"form","id":"drlgrqckeBor"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@title ### 4. Plotting the RMSF of our short simulation\n","import matplotlib.pyplot as plt\n","plt.style.use('seaborn-whitegrid')\n","import numpy as np\n","\n","#Reading the text file containing this information\n","data = np.loadtxt('rmsf.xvg')\n","\n","plt.title('Ca-RMSF during 0.1 ns production MD')\n","plt.xlabel('residue number')\n","plt.ylabel('Ca-RMSF (nm)')\n","plt.plot(data[:,0], data[:,1], linestyle='solid', linewidth='2', color='red') \n","plt.show()"],"metadata":{"cellView":"form","id":"Uqj00YMngNEW"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"5aPBlo6fp5xn"},"source":["To finalize, we will visualize our simulation. For this, we will use the `trjconv` module from GROMACS to remove the solvent from our system and convert our trajectory into a PDB file.\n","\n","Given that our protein may have drifted from the center of the box and reached its edges, we will also take the opportunity to recenter the periodic box on our protein such that its atoms do not stretch through the edges of the periodic boundary conditions."]},{"cell_type":"code","metadata":{"id":"OdH62ynFqRsc","cellView":"form"},"source":["#@title ### 3. Removing the solvent from our MD simulation\n","%%bash\n","source /content/gromacs/bin/GMXRC\n","\n","#This is a trick to provide interactive options to gmx\n","echo \"Protein\" > options\n","echo \" \" >> options\n","echo \"Protein\" >> options\n","echo \" \" >> options\n","\n","#Using trjconv to extract only the protein atoms from the simulation trajectory\n","#and also recenter the protein if its atoms crossed the periodic boundaries\n","gmx trjconv -s md_1.tpr -f md_1.xtc -o md_traj.pdb -pbc mol -center -quiet < options"],"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@markdown Now, you can download this new PDB file and load it onto [**NGLviewer**](http://nglviewer.org/ngl/) as a **trajectory** PDB file to visualize the protein motions explored within this short simulation time.\n","files.download(f\"md_traj.pdb\")"],"metadata":{"cellView":"form","id":"eJwUEKGbbNgl"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# Part V - Backing up your files\n","\n","1. If you want to download your produced files, execute the code below. A compressed .tar.gz file will be generated and automatically downloaded into your computer (unless you have an ad-blocker, for which you will have to manually download it)."],"metadata":{"id":"eMPMYnpF8BDC"}},{"cell_type":"code","source":["#Compressing all files into a .tar.gz file\n","!tar -czf D2-tutorial-03.tar.gz *"],"metadata":{"id":"8UegLlFG8oZD"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["from google.colab import files\n","files.download(\"/content/D2-tutorial-01.tar.gz\")"],"metadata":{"id":"UC4GuswE8qxW"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["2. Alternatively, you can transfer the files directly to your Google Drive as shown below:"],"metadata":{"id":"IoHmyfPC8sPn"}},{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive')"],"metadata":{"id":"e4rekcvQ8t6H"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["import os\n","import shutil\n","from pathlib import Path \n","backup = Path(\"/content/drive/MyDrive/saocarlos2022/\")\n","if os.path.exists(backup):\n","  print(\"Sao Carlos Workshop 2022 - Backup folder already exists\")\n","if not os.path.exists(backup):\n","  os.mkdir(backup)\n","  print(\"Sao Carlos Workshop 2022 - Backup folder did not exists and was succesfully created\")\n","\n","#Backing up\n","shutil.copy(str('/content/D2-tutorial-03.tar.gz'), str(backup/'D2-tutorial-03.tar.gz'))\n","print(\"Day 2 - Tutorial 3 files successfully backed up!\")"],"metadata":{"id":"oFNJDk3k8uaL"},"execution_count":null,"outputs":[]}]}